app/models/patron_request.rb

```diff
@@ -46,16 +46,63 @@ def default_service_point_code
     service_points.first || Settings.folio.default_service_point
   end
 
-  private
+  def mediateable?
+    items_in_location.all?(&:mediateable?)
+  end
+
+  def requires_needed_date?
+    return false if origin_location_code == 'PAGE-MP' || origin_location_code == 'SAL3-PAGE-MP'
+
+    true
+  end
+
+  def use_in_library?
+    items_in_location.all? { |item| item.effective_location.discovery_display_name.include?('For use in') }
+  end
+
+  def single_location_display
+    Folio::Types.service_points.find_by(code: pickup_destinations.first)
+  end
+
+  def single_location_label
+    return 'Must be used in library' if mediateable? || use_in_library?
+
+    'Will be delivered to'
+  end
+
+  def destination_library_code
+    find_location_library(single_location_display.code) if single_location_display
+  end
+
+  def find_location_library(lib_code)
+    Folio::Types.libraries.find_by(code: lib_code)&.code
+  end
+
+  def earliest_delivery_estimate
+    paging_info = PagingSchedule.for(self)
+    { 'date' => Time.zone.today + paging_info.days_later, 'time' => paging_info.will_arrive_text }
+  rescue StandardError
+    nil
+  end
+
+  def date_picker_date
+    return earliest_delivery_estimate['date'] if earliest_delivery_estimate
+
+    Time.zone.today
+  end
 
   def folio_location
     @folio_location ||= Folio::Types.locations.find_by(code: origin_location_code) || items_in_location.first&.permanent_location
   end
 
   def origin_library_code
-    folio_location&.library&.code
+    return folio_location&.library&.code if folio_location&.library
+
+    Folio::Types.libraries.find_by(id: folio_location.library_id)&.code
```

Comment:
What scenarios do we have the folio location without a library code? Is this an artifact of testing or something else?

Reply:
This is a specific example that I ran into. It was either http://localhost:3000/patron_requests/new?instance_hrid=8834398&origin_location_code=ART-LOCKED-SMALL or http://localhost:3000/patron_requests/new?instance_hrid=2892728&origin_location_code=SAL3-PAGE-MP. It did not have a library on it's record but it did have a library_id. It was very strange.

Reply:
I was the first one. For some reason this with is returned for folio_location for ART-LOCKED-SMALL

<data Folio::Location id="01e91968-2e52-4733-bbcc-b4641560411d", campus=nil, library=nil, library_id="2926c7ed-ed36-4385-8503-58cd6cee73e8", institution=#<data Folio::Institution id="8d433cdd-4e8f-4dc1-aa24-8a4ddb7dc929">, code="ART-LOCKED-SMALL", discovery_display_name="Locked stacks (Small)", name="Art Locked Stacks Small", primary_service_point_id="77cd12ac-2de8-4d13-99a0-f6b3b4f4bdca", details={"pageMediationGroupKey"=>"ART", "pageServicePointCodes"=>"ART"}>